proposed/2022/package-status-indicators.md

```diff
@@ -0,0 +1,120 @@
+## Explanation
+
+### Functional explanation
+
+<!-- Explain the proposal as if it were already implemented and you're teaching it to another person. -->
+<!-- Introduce new concepts, functional designs with real life examples, and low-fidelity mockups or  pseudocode to show how this proposal would look. -->
+
+When a developer is working within Visual Studio, they may regularly check their "Dependencies" node within the solution explorer. When expanded, a "Packages" node is available to show the hierarchy of top-level and transitive packages included in a project.
+
+NuGet uses three main words to describe these status of a package, the TL;DR are:
+
+- **Deprecated** - Please don't use this for a disclosed reason and please use a suggested alternative(if any).
+- **Vulnerable** - Please don't use this due to a known vulnerability and update/replace/remove/do nothing accordingly.
+- **Outdated** - Please consider using a newer version.
+
+Building upon Tim's initial mockup, there should be a UI affordance that demonstrates the truth of each package's status in the solution explorer representing these three words.
+
+![](../../meta/resources/PackageStatusIndicators/package-status-indicators.png)
+
+For the sake of not annoying developers, we should only bubble up a warning affordance to the top-level node (Packages and Dependencies) if there is a known top-level vulnerability or deprecated package in their package graph.
+
+Outdated packages would still show the indicators if the packages node is expanded however.
+
+Currently, NuGet only uses two icons to display status indicators:
+
+- A warning triangle (vulnerability and deprecation).
+![](../../meta/resources/PackageStatusIndicators/StatusWarning.png)
+- An up arrow (outdated).
+![](../../meta/resources/PackageStatusIndicators/StatusUpdateAvailable.png)
+
+Hovering over the package and icon will display the respective information for the status.
+
+- Deprecated - "This package has been deprecated as it is `<deprecation category>`."
+- Vulnerable - "This package has at least one vulnerability with `<severity category>` severity."
+- Outdated - "This package has version `<latest version>` available.
+
+When right clicking a package, a developer should be able to get to the "Manager NuGet Packages..." context menu action like they can with the "Packages" and "Dependencies" nodes.
+
+Because this feature may need some experimentation to get the right signal from the noise and the potential for exponential transitive-level vulnerabilities, developers will need a way to opt-out from the experience as a whole, and options to opt-in to varying levels of information.
+
+This may need to include a new options panel inside the `Projects and Solutions > SDK-Style Projects` regarding the Package Settings where there's a top-level checkbox for the feature, and either a drop-down indicating a package status level or multiple checkboxes for each status mentioned above.
+
+<!-- ### Technical explanation -->
```

Comment:
I'd love to see a technical explanation that focuses on the feasibility of some of these asks. That might allow to define the stages for this work.

Vulnerabilities are going to always be up to date. 

What happens with deprecation & update, which are expensive to calculate for example.

Things I'm curious about. 

- How often are they going to be done? How frequent should the refresh be?
- Any correctness concerns, given that this simply can't happen at every restore, it'd be an expensive perf hit right now.

Reply:
Data about deprecations/vulnerabilities/updates/etc could be handled async relative to the currently project data flow. In VS, the total set of packages in the solution could be gathered and this information queried. Once available, it could be pushed to projects, which could then update project state. 

If the data was cached, then solution-load could efficiently load the data during solution load so that we show that state immediately.

The component that manages that state could also be responsible for keeping it updated via whatever means makes sense, such as polling a remote server.

One area I'm curious about is vulnerable transitive dependencies. The project references A which references B, and B is vulnerable. We would want to show A as vulnerable too. Therefore this component would need to model dependencies between packages and push any such state back up the graph to referencing packages, recursively.

Reply:
> One area I'm curious about is vulnerable transitive dependencies. The project references A which references B, and B is vulnerable. We would want to show A as vulnerable too. Therefore this component would need to model dependencies between packages and push any such state back up the graph to referencing packages, recursively.

Transitive dependencies is definitely an area where we needs lots of work. 
By default the .NET 8 SDK will report on directs, but transitive dependencies can have really *large* graphs.

Reply:
Given there are commonly many more indirect dependencies than direct dependencies, I wouldn't be surprised if the majority of exposure to vulnerable packages occurs through indirect dependencies.